kroxylicious-authorizer-providers/kroxylicious-authorizer-acl/src/main/java/io/kroxylicious/authorizer/provider/acl/AclAuthorizer.java

```diff
@@ -0,0 +1,466 @@
+                                                                   Class<? extends Principal> principalType,
+                                                                   TypeNameMap.Predicate principalPredicate,
+                                                                   @Nullable String principalName,
+                                                                   Class<O> opType,
+                                                                   Pred resourceNamePredicate,
+                                                                   @Nullable String resourceName,
+                                                                   Set<O> operations) {
+        var es = EnumSet.copyOf(operations);
+        for (var op : es) {
+            es.addAll(op.implies());
+        }
+        PrincipalGrants compute = allowPerPrincipal.compute(principalType, principalName, principalPredicate,
+                g -> {
+                    if (g == null) {
+                        return new PrincipalGrants(resourceNamePredicate == Pred.MATCH ? null : new TypeNameMap<>(),
+                                resourceNamePredicate == Pred.MATCH ? new TypePatternMatch() : null);
+                    }
+                    else if (g.patternMatch() == null && resourceNamePredicate == Pred.MATCH) {
+                        return new PrincipalGrants(g.nameMatches(), new TypePatternMatch());
+                    }
+                    else if (g.nameMatches() == null && resourceNamePredicate != Pred.MATCH) {
+                        return new PrincipalGrants(new TypeNameMap<>(), g.patternMatch());
+                    }
+                    return g;
+                });
+
+        if (resourceNamePredicate == Pred.MATCH) {
+            Objects.requireNonNull(compute.patternMatch()).compute(opType, Pattern.compile(Objects.requireNonNull(resourceName)), operations);
+        }
+        else {
+            Objects.requireNonNull(compute.nameMatches()).compute(opType, resourceName, Objects.requireNonNull(resourceNamePredicate.toNameMatchPredicate()),
+                    v -> {
+                        if (v == null) {
+                            return es;
+                        }
+                        es.addAll((EnumSet) v);
+                        return es;
+                    });
+        }
+    }
+
+    private static @Nullable Decision authorizeInternal(
+                                                        TypeNameMap<Principal, PrincipalGrants> allowPerPrincipal,
+                                                        Subject subject,
+                                                        Action action,
+                                                        Decision whenFound,
+                                                        @Nullable Decision whenNotFound) {
+        assert (whenFound != whenNotFound);
+        for (var p : subject.principals()) {
```

Comment:
What if we ALLOW principal X and DENY principal Y? It looks like the first match would win, and depend on the order in the subject.